File: LeetCode/LeetCode/OtherActivities/AlgorithmStudyPlan/Algorithm2/Problem34.cs
namespace LeetCode.OtherActivities.AlgorithmStudyPlan.Algorithm2
{
    public class Problem34
    {
        public int[] SearchRange(int[] nums, int target)
        {
            var result = new int[2];

            var left = 0;
            var right = nums.Length - 1;

            while (left < right)
            {
                var mid = (right + left) / 2;

                if (nums[mid] >= target)
                {
                    right = mid;
                }
                else
                {
                    left = mid + 1;
                }
            }

            if (nums.Length > 0 && nums[left] == target)
                result[0] = left;
            else
            {
                result[0] = -1;
                result[1] = -1;
                return result;
            }

            left = 0;
            right = nums.Length - 1;
            while (left < right)
            {
                var mid = (right + left) / 2;

                if (nums[mid] <= target)
                {
                    if (left == mid)
                    {
                        if (nums[mid + 1] == target)
                            left++;

                        break;
                    }

                    left = mid;
                }
                else
                {
                    right = mid - 1;
                }
            }

            result[1] = left;

            return result;
        }
    }
}

File: leetcode_rust/src/problems/problem_34.rs
pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut res = vec![-1, -1];

    if nums.len() == 0 {
        return res;
    }

    let mut left = 0usize;
    let mut right = nums.len() - 1;

    while left < right {
        let mid = (right + left) / 2;
        if nums[mid] < target {
            left = mid + 1
        } else {
            right = mid;
        }
    }

    if nums[left] == target {
        res[0] = left as i32;
    } else {
        return res;
    }

    left = 0;
    right = nums.len() - 1;

    while left < right {
        let mid = (right + left) / 2;

        if nums[mid] > target {
            right = mid - 1;
        } else {
            if left == mid {
                if nums[mid + 1] == target {
                    left += 1;
                }
                break;
            }
            left = mid;
        }
    }

    res[1] = left as i32;

    res
}
